contracts/cw3-flex-multisig/src/contract.rs

```diff
@@ -0,0 +1,993 @@
+        .take(limit)
+        .map(|p| map_proposal(&env.block, p))
+        .collect();
+
+    Ok(ProposalListResponse { proposals: props? })
+}
+
+fn reverse_proposals(
+    deps: Deps,
+    env: Env,
+    start_before: Option<u64>,
+    limit: Option<u32>,
+) -> StdResult<ProposalListResponse> {
+    let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+    let end = start_before.map(Bound::exclusive_int);
+    let props: StdResult<Vec<_>> = PROPOSALS
+        .range(deps.storage, None, end, Order::Descending)
+        .take(limit)
+        .map(|p| map_proposal(&env.block, p))
+        .collect();
+
+    Ok(ProposalListResponse { proposals: props? })
+}
+
+fn map_proposal(
+    block: &BlockInfo,
+    item: StdResult<(Vec<u8>, Proposal)>,
+) -> StdResult<ProposalResponse> {
+    let (key, prop) = item?;
+    let status = prop.current_status(block);
+    Ok(ProposalResponse {
+        id: parse_id(&key)?,
+        title: prop.title,
+        description: prop.description,
+        msgs: prop.msgs,
+        expires: prop.expires,
+        status,
+    })
+}
+
+fn query_vote(deps: Deps, proposal_id: u64, voter: HumanAddr) -> StdResult<VoteResponse> {
+    let voter_raw = deps.api.canonical_address(&voter)?;
+    let prop = BALLOTS.may_load(deps.storage, (proposal_id.into(), &voter_raw))?;
+    let vote = prop.map(|b| b.vote);
+    Ok(VoteResponse { vote })
+}
+
+fn list_votes(
+    deps: Deps,
+    proposal_id: u64,
+    start_after: Option<HumanAddr>,
+    limit: Option<u32>,
+) -> StdResult<VoteListResponse> {
+    let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+    let canon = maybe_canonical(deps.api, start_after)?;
+    let start = canon.map(Bound::exclusive);
+
+    let api = &deps.api;
+    let votes: StdResult<Vec<_>> = BALLOTS
+        .prefix(proposal_id.into())
+        .range(deps.storage, start, None, Order::Ascending)
+        .take(limit)
+        .map(|item| {
+            let (key, ballot) = item?;
+            Ok(VoteInfo {
+                voter: api.human_address(&CanonicalAddr::from(key))?,
+                vote: ballot.vote,
+                weight: ballot.weight,
+            })
+        })
+        .collect();
+
+    Ok(VoteListResponse { votes: votes? })
+}
+
+fn query_voter(deps: Deps, voter: HumanAddr) -> StdResult<VoterResponse> {
+    let cfg = CONFIG.load(deps.storage)?;
+    let voter_raw = deps.api.canonical_address(&voter)?;
+    let weight = cfg
+        .group
+        .is_member(&deps.querier, &voter_raw)?
+        .unwrap_or_default();
+
+    Ok(VoterResponse {
+        addr: voter,
+        weight,
+    })
+}
+
+fn list_voters(
+    deps: Deps,
+    start_after: Option<HumanAddr>,
+    limit: Option<u32>,
+) -> StdResult<VoterListResponse> {
+    let cfg = CONFIG.load(deps.storage)?;
+    let voters = cfg
+        .group
+        .list_members(&deps.querier, start_after, limit)?
+        .into_iter()
+        .map(|member| VoterResponse {
```

Comment:
Probably not worth it to simplify, but `Member` and `VoterResponse` are basically the same.

Reply:
They are the same. They are defined in two different specs, cw3 and cw4.

Maybe I should combine them into one shared type (and then make an explicit dependency between the two specs). This can be another issue, and if you think it simplifies the code, happy if you make such a PR. I wonder if the Rust compiler will reduce this `.map()` into a no-op, realizing the two types have the same byte layout.

Reply:
I guess we would need to dissassemble to know.

Regarding combining the types, I'll create the issue, and work on a PR to see if it's worth it.

Reply:
See #151.